File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/DatabaseRequestHandler.swift
//
//  DatabaseRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Raghu Sarangapani on 12/20/17.
//  Copyright © 2017 Raghu Sarangapani. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift


public class DatabaseRequestHandler {
    public static let VOID: String? = nil
    fileprivate var _pushPullReplListener:NSObjectProtocol?
    
    public func handleRequest(method: String, args: Args) throws -> Any? {
        switch method {

        //////////////
        // Database //
        //////////////
    
        case "database_create":
            let name: String! = args.get(name: "name")
            let dbConfig: DatabaseConfiguration? = args.get(name: "config")
            do {
                if dbConfig != nil {
                    return try Database(name: name!, config: dbConfig!)
                } else {
                    return try Database(name: name!)
                }
            } catch {
                return error.localizedDescription
            }
            
        case "database_close":
            let database: Database = args.get(name:"database")!
            try database.close()

        case "database_getPath":
            let database: Database = (args.get(name:"database"))!

            return database.path

        case "database_delete":
            let database: Database = (args.get(name:"database"))!
            let document: Document = (args.get(name:"document"))!

            do {
                try database.deleteDocument(document)
            } catch {
                return error
            }
            
        case "database_deleteBulkDocs":
            let database: Database = args.get(name:"database")!
            let doc_ids: Array<String> = args.get(name: "doc_ids")!

            try database.inBatch {
                for id in doc_ids {
                    let document: Document = database.document(withID: id)!
                    do {
                        try! database.deleteDocument(document)
                    }
                }
            }

        case "database_deleteDB":
            let database: Database = args.get(name:"database")!
            
            try database.delete()

        case "database_exists":
            let name: String = args.get(name:"name")!
            let directory: String? = args.get(name:"directory")!
            return Database.exists(withName: name, inDirectory: directory)

        case "database_deleteIndex":
            let database: Database = (args.get(name:"database"))!
            let name: String = (args.get(name: "name"))!
            
            try database.deleteIndex(forName: name)

        case "database_getName":
            let database: Database = args.get(name:"database")!

            return database.name

        case "database_getDocument":
            let database: Database = (args.get(name:"database"))!
            let id: String? = args.get(name: "id")
            return database.document(withID: id!)
            
        case "database_save":
            let database: Database = (args.get(name:"database"))!
            let document: MutableDocument = args.get(name:"document")!
            try! database.saveDocument(document)
            
        case "database_saveWithConcurrency":
            let database: Database = (args.get(name:"database"))!
            let document: MutableDocument = args.get(name:"document")!
            let concurrencyControlType : String? = args.get(name:"concurrencyControlType")!
            var concurrencyType = ConcurrencyControl.lastWriteWins
            
            if let type = concurrencyControlType {
                if type == "failOnConflict" {
                    concurrencyType = .failOnConflict
                } else {
                    concurrencyType = .lastWriteWins
                }
            }
            try! database.saveDocument(document, concurrencyControl: concurrencyType)

        case "database_deleteWithConcurrency":
            let database: Database = (args.get(name:"database"))!
            let document: Document = (args.get(name:"document"))!
            let concurrencyControlType : String? = args.get(name:"concurrencyControlType")!
            var concurrencyType = ConcurrencyControl.lastWriteWins
            
            if let type = concurrencyControlType {
                if type == "failOnConflict" {
                    concurrencyType = .failOnConflict
                } else {
                    concurrencyType = .lastWriteWins
                }
            }
            do {
                try database.deleteDocument(document, concurrencyControl: concurrencyType)
            } catch {
                return error
            }

            
        case "database_purge":
            let database: Database = (args.get(name:"database"))!
            let document: MutableDocument = args.get(name:"document")!
            try! database.purgeDocument(document)
            
        case "database_contains":
            let database: Database = (args.get(name:"database"))!
            let id: String = (args.get(name: "id"))!

            return database.document(withID: id) != nil

        case "database_getCount":
            let database: Database = (args.get(name:"database"))!
            return database.count

        case "database_compact":
            let database: Database = (args.get(name:"database"))!
            return try! database.compact()
            
        case "database_addChangeListener":
            let database: Database = (args.get(name:"database"))!
            var token: ListenerToken? = nil
            let docId: String? = (args.get(name:"docId"))
            if (docId != nil) {
                let changeListener = MyDocumentChangeListener()
                token = database.addDocumentChangeListener(withID: docId!, listener: changeListener.listener)
            } else {
                let changeListener = MyDatabaseChangeListener()
                token = database.addChangeListener(changeListener.listener)
            }
            
            return token

        case "database_removeChangeListener":
            let database: Database = (args.get(name:"database"))!
            let changeListener: ListenerToken = (args.get(name: "changeListener"))!

            database.removeChangeListener(withToken: changeListener)

        case "database_databaseChangeListenerChangesCount":
            let changeListener: MyDatabaseChangeListener = (args.get(name: "changeListener"))!

            return changeListener.getChanges().count

        case "database_databaseChangeListenerGetChange":
            let changeListener: MyDatabaseChangeListener = (args.get(name: "changeListener"))!
            let index: Int = (args.get(name: "index"))!

            return changeListener.getChanges()[index]

        case "database_databaseChangeListenerGetChanges":
            let changeListener: MyDatabaseChangeListener = (args.get(name: "changeListener"))!
            return changeListener.getChanges()
            
        case "database_databaseChangeGetDocumentId":
            let change: DatabaseChange = (args.get(name: "change"))!

            return change.documentIDs
            
        case "database_deleteDBbyName":
            let name: String = args.get(name:"name")!
            let directory: String? = args.get(name:"directory")!
            if let directory = directory {
                return try Database.delete(withName: name, inDirectory: directory)
            } else {
                return try Database.delete(withName: name)
            }

        case "database_saveDocuments":
            let database: Database = args.get(name:"database")!
            let documents: Dictionary<String, Dictionary<String, Any>> = args.get(name: "documents")!
            try database.inBatch {
                for doc in documents {
                    let id = doc.key
                    let data: Dictionary<String, Any> = doc.value
                    let document = MutableDocument(id: id, data: data)
                    try! database.saveDocument(document)
                    
                }
            }
 
        case "database_updateDocuments":
            let database: Database = args.get(name:"database")!
            let documents: Dictionary<String, Dictionary<String, Any>> = args.get(name: "documents")!
            try database.inBatch {
                for doc in documents {
                    let id = doc.key
                    let data: Dictionary<String, Any> = doc.value
                    let updated_doc = database.document(withID: id)!.toMutable()
                    updated_doc.setData(data)
                    try database.saveDocument(updated_doc)
                }
            }
            
        case "database_updateDocument":
            
            let database: Database = (args.get(name:"database"))!
            let data: Dictionary<String, Any> = args.get(name: "data")!
            let docId: String = args.get(name: "id")!
            let updated_doc = database.document(withID: docId)!.toMutable()
            updated_doc.setData(data)
            try! database.saveDocument(updated_doc)
            
            
        case "database_getDocIds":
            let database: Database = args.get(name:"database")!
            let limit: Int = args.get(name:"limit")!
            let offset: Int = args.get(name:"offset")!
            let query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.database(database))
                .limit(Expression.int(limit), offset:Expression.int(offset))

            var result: [String] = []
            do {
                for row in try query.execute() {
                    result.append(row.string(forKey: "id")!)
                }
            }

            return result

        case "database_getDocuments":
            let database: Database = args.get(name:"database")!
            let ids: [String] = args.get(name:"ids")!
            var documents = [String: [String: Any]]()

            for id in ids {
                let document: Document = database.document(withID: id)!
                
                var dict = document.toDictionary()
                for (key, value) in dict {
                    if let list = value as? Dictionary<String, Blob> {
                        var item = Dictionary<String, Any>()
                        for (k, v) in list {
                            item[k] = v.properties
                        }
                        dict[key] = item
                    }
                }
                documents[id] = dict
            }

            return documents
          
        #if COUCHBASE_ENTERPRISE
        case "database_changeEncryptionKey":
            let database: Database = args.get(name:"database")!
            let password: String? = args.get(name:"password")!
            let encryptionKey: EncryptionKey?
            if password == "nil"{
                encryptionKey = nil
            }
            else{
                encryptionKey = EncryptionKey.password(password!)
            }
            do {
                return try database.changeEncryptionKey(encryptionKey)
            } catch {
                print("Got error setting encryption key \(error)")
                return error.localizedDescription
            }
        #endif
        case "database_copy":
            let dbName: String! = args.get(name: "dbName")
            let dbConfig: DatabaseConfiguration? = args.get(name: "dbConfig")
            let dbPath: String! = args.get(name: "dbPath")
            try! Database.copy(fromPath: dbPath, toDatabase: dbName, withConfig: dbConfig)

        case "database_getPreBuiltDb":
            let dbPath: NSString! = args.get(name: "dbPath")
            let path: String! = Bundle(for: type(of:self)).path(forResource: dbPath.deletingPathExtension, ofType: dbPath.pathExtension)
            return path

        case "database_queryAllDocuments":
            let database: Database = args.get(name:"database")!
            let searchQuery = QueryBuilder
                .select(SelectResult.all())
                .from(DataSource.database(database))

            return searchQuery
        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
        return DatabaseRequestHandler.VOID
    }
}

class MyDatabaseChangeListener  {
    var changes: [DatabaseChange] = []
    
    lazy var listener: (DatabaseChange) -> Void = { (change: DatabaseChange) in
        self.changes.append(change)
    }
    
    public func getChanges() -> [DatabaseChange] {
        return changes
    }
}

File: CBLClient/Apps/CBLTestServer-Android/app/src/main/java/com/couchbase/CouchbaseLiteServ/server/RequestHandler/DatabaseRequestHandler.java
package com.couchbase.CouchbaseLiteServ.server.RequestHandler;

import android.content.Context;
import android.util.Log;

import com.couchbase.CouchbaseLiteServ.server.Args;
import com.couchbase.lite.ConcurrencyControl;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseChange;
import com.couchbase.lite.DatabaseChangeListener;
import com.couchbase.lite.DatabaseConfiguration;
import com.couchbase.lite.Document;
import com.couchbase.CouchbaseLiteServ.MainActivity;
import com.couchbase.lite.Expression;
import com.couchbase.lite.ListenerToken;
import com.couchbase.lite.Meta;
import com.couchbase.lite.MutableDocument;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Result;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.SelectResult;
import com.couchbase.lite.EncryptionKey;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DatabaseRequestHandler {
    /* ------------ */
    /* - Database - */
    /* ------------ */

    public Database create(Args args) throws CouchbaseLiteException {
        String name = args.get("name");
        DatabaseConfiguration config = args.get("config");
        if (config == null) {
            Context context = MainActivity.getAppContext();
            config = new DatabaseConfiguration(context);
        }
        return new Database(name, config);
    }

    public ListenerToken addChangeListener(Args args) {
        Database database = args.get("database");
        MyDatabaseChangeListener changeListener = new MyDatabaseChangeListener();
        return database.addChangeListener(changeListener);
    }

    public void removeChangeListener(Args args) {
        Database database = args.get("database");
        ListenerToken changeListener = args.get("changeListener");
        database.removeChangeListener(changeListener);
    }

    public Database create(String name) throws CouchbaseLiteException {
        return new Database(name, null);
    }

    public long getCount(Args args) {
        Database database = args.get("database");
        return database.getCount();
    }

    public void close(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        database.close();
    }

    public void compact(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        database.compact();
    }

    public String getPath(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        return database.getPath();
    }

    public String getName(Args args) {
        Database database = args.get("database");
        return database.getName();
    }

    public Document getDocument(Args args) {
        Database database = args.get("database");
        String id = args.get("id");
        return database.getDocument(id);
    }

    public List<String> getIndexes(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        return database.getIndexes();
    }

    public Map<String, Map<String, Object>> getDocuments(Args args) {
        Database database = args.get("database");
        List<String> ids = args.get("ids");
        Map<String, Map<String, Object>> documents = new HashMap<>();
        for (String id : ids) {
            Document document = database.getDocument(id);
            if (document != null) {
                documents.put(id, document.toMap());
            }
        }
        return documents;
    }

    public void updateDocument(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        MutableDocument document = args.get("document");
        String id = document.getId();
        Map<String, Object> data = (Map<String, Object>) document.getValue(id);
        MutableDocument updateDoc = database.getDocument(id).toMutable();
        updateDoc.setData(data);
        database.save(updateDoc);
    }

    public void updateDocuments(Args args) throws CouchbaseLiteException {
        final Database database = args.get("database");
        final Map<String, Map<String, Object>> documents = args.get("documents");
        database.inBatch(new Runnable() {
            @Override
            public void run() {
                for (Map.Entry<String, Map<String, Object>> entry : documents.entrySet()) {
                    String id = entry.getKey();
                    Map<String, Object> data = entry.getValue();
                    MutableDocument updatedDoc = database.getDocument(id).toMutable();
                    updatedDoc.setData(data);
                    try {
                        database.save(updatedDoc);
                    } catch (CouchbaseLiteException e) {
                        e.printStackTrace();
                    }
                }
            }
        });
    }

    public void purge(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        MutableDocument document = args.get("document");
        database.purge(document);
   }

   public void saveDocuments(Args args) throws CouchbaseLiteException {
        final Database database = args.get("database");
        final Map<String, Map<String, Object>> documents = args.get("documents");

        database.inBatch(new Runnable() {
            @Override
            public void run() {
                for (Map.Entry<String, Map<String, Object>> entry : documents.entrySet()) {
                    String id = entry.getKey();
                    Map<String, Object> data = entry.getValue();
                    MutableDocument document = new MutableDocument(id, data);
                    try {
                        database.save(document);
                    } catch (CouchbaseLiteException e) {
                        e.printStackTrace();
                    }
                }
            }
        });
   }

    public void save(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        MutableDocument document = args.get("document");
        database.save(document);
    }

    public void saveWithConcurrency(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        MutableDocument document = args.get("document");
        String concurrencyControlType = args.get("concurrencyControlType");
        ConcurrencyControl concurrencyType;
        if (concurrencyControlType == null)
        {
            concurrencyType = ConcurrencyControl.LAST_WRITE_WINS;
        }
        if(concurrencyControlType.equals("failOnConflict"))
            concurrencyType = ConcurrencyControl.FAIL_ON_CONFLICT;
        else
            concurrencyType = ConcurrencyControl.LAST_WRITE_WINS;
        database.save(document, concurrencyType);
    }

    public void delete(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        Document document = args.get("document");
        database.delete(document);
   }

    public void deleteWithConcurrency(Args args) throws CouchbaseLiteException {
      Database database = args.get("database");
      Document document = args.get("document");
      String concurrencyControlType = args.get("concurrencyControlType");
      ConcurrencyControl concurrencyType;
      if (concurrencyControlType == null)
      {
        concurrencyType = ConcurrencyControl.LAST_WRITE_WINS;
      }
      if(concurrencyControlType.equals("failOnConflict"))
        concurrencyType = ConcurrencyControl.FAIL_ON_CONFLICT;
      else
        concurrencyType = ConcurrencyControl.LAST_WRITE_WINS;
      database.delete(document, concurrencyType);
    }

    public void deleteDB(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        try {
            database.delete();
        }catch (CouchbaseLiteException ex){
            Log.e("deleteDB", "deleteDB() ERROR !!!!!!");
            ex.printStackTrace();
        }
        System.out.println("database delete ");
   }

    public void changeEncryptionKey(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        String password = args.get("password");
        EncryptionKey encryptionKey;
        if(password.equals("nil"))
          encryptionKey = null;
        else
          encryptionKey = new EncryptionKey(password);
        database.changeEncryptionKey(encryptionKey);
    }

   public void deleteDbByName(Args args) throws CouchbaseLiteException {
        String dbName = args.get("dbName");
        File directory = args.get("directory");
        Database.delete(dbName, directory.getParentFile());
   }

    public boolean exists(Args args){
        String name = args.get("name");
        File directory = new File(args.get("directory").toString());
        return Database.exists(name, directory);
    }

    public void deleteBulkDocs(Args args) throws CouchbaseLiteException {
        final Database db = args.get("database");
        final List<String> docIds = args.get("doc_ids");
        db.inBatch(new Runnable() {
            @Override
            public void run() {
                for(String id : docIds) {
                    Document document = db.getDocument(id);
                    try {
                        db.delete(document);
                    } catch (CouchbaseLiteException e) {
                        e.printStackTrace();
                    }
                }
                }
        });

    }
/*    method dropped in DB022
    public boolean contains(Args args) {
       Database database = args.get("database");
       String id = args.get("id");
       return database.contains(id);
    }*/

    public List<String> getDocIds(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");
        int limit = args.get("limit");
        int offset = args.get("offset");
        Query query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.database(database))
                .limit(Expression.intValue(limit), Expression.intValue(offset));
        List<String> result = new ArrayList<String>();
        ResultSet results = query.execute();
        for (Result row : results){

            result.add(row.getString("id"));
        }
        Collections.sort(result);
        return result;

    }

//    public void addChangeListener(Args args) {
//        Database database = args.get("database");
//        if (args.contain("docId")) {
//            String docId = args.get("docId");
//            MyDocumentChangeListener changeListener = new MyDocumentChangeListener();
//            database.addChangeListener(docId, changeListener);
//        } else {
//            MyDatabaseChangeListener changeListener = new MyDatabaseChangeListener();
//            database.addChangeListener(changeListener);
//        }
//    }

//    public void removeChangeListener(Args args) {
//        Database database = args.get("database");
//        if (args.contain("docId")){
//        String docId = args.get("docId");
//        DocumentChangeListener changeListener = args.get("changeListener");
//        database.removeChangeListener(docId, changeListener);
//        } else{
//            DatabaseChangeListener changeListener = args.get("changeListener");
//            database.removeChangeListener(changeListener);
//        }
//    }

    public int databaseChangeListenerChangesCount(Args args) {
        MyDatabaseChangeListener changeListener = args.get("changeListener");
        return changeListener.getChanges().size();
    }

    public DatabaseChange databaseChangeListenerGetChange(Args args) {
        MyDatabaseChangeListener changeListener = args.get("changeListener");
        int index = args.get("index");
        return changeListener.getChanges().get(index);
    }

    public Database changeGetDatabase(Args args){
        DatabaseChange change = args.get("change");
        return change.getDatabase();
    }

    public List<String> changeGetDocumentId(Args args) {
        DatabaseChange change = args.get("change");
        return change.getDocumentIDs();
    }

}

class MyDatabaseChangeListener implements DatabaseChangeListener{
    private List<DatabaseChange> changes;

    public List<DatabaseChange> getChanges(){
       return changes;
    }

    @Override
    public void changed(DatabaseChange change) {
        changes.add(change);
    }
}